### ISSUE TITLE
JWTAuth::make() and JWTAuth::encode() calls produce previously invalidated token.

### ISSUE THREAD

**Issue by Cheezykins** [tymondesigns/jwt-auth#688]
Hi,
I'm not using users within the laravel app itself, but rather an upstream authentication provider.

because of this I have the following middleware which invalidates and regenerates each token on use, this is to allow it to ping the upstream provider to keep the session alive.

In my route middleware I am getting the payload of a token then invalidating it with the code:

```php
        try {
            $auth = JWTAuth::parseToken();
            $claims = JWTAuth::getPayLoad()->toArray();
            JWTAuth::invalidate();
        } catch (TokenExpiredException $e) {
            abort(403, 'Your token has expired');
        } catch (JWTException $e) {
            abort(403, 'Invalid token');
        }

        $token = User::ping($claims);

        $response = $next($request);
        $response->headers->set('Authorization', 'Bearer ' . $token->get());

        return $response;
```

Then within the ping call I am passing some of the claims through to create a new token with the code:

```php
    public static function ping($claims)
    {
        $now = new Carbon('now', 'UTC');

        if ($now->getTimestamp() > $claims['fxp']) {
            // upstream authentication session expired
            abort(403, 'your token has expired');
        }

        if (($claims['fxp'] - $now->getTimestamp()) < 180) {
            // upstream authentication session expires soon
            // ping it.
            $upstream->ping($claims['sub']);
            $claims['fxp'] = $upstream->getSessionExpiry();
        }

        $newClaims = [
            'sub' => $claims['sub'],
            'fxp' => $claims['fxp'],
            'iss' => env('API_SUBTYPE')
        ]

        $payLoad = JWTFactory::make($newClaims);
        return JWTAuth::encode($payLoad);
    }
```

The problem is, the token generated by JWTAuth::encode($payLoad) in my ping method sometimes returns the same token that is invalidated above in my middleware, even though the payload should have changed (exp and jti claims)

Here's a gist containing some logging I added to the code:

https://gist.github.com/Cheezykins/2e2d04b3f14f1e675b5ab6c80a747e97

Is this a bug or am I doing something fundamentally wrong?

**Comment by Cheezykins** [comment]
I am more certain that this is a bug now.

I created a new service provider that overrode the manager and extended the JWTManager::refresh() method with my logic, then modelled my middleware after the included RefreshToken middleware.

Here's the code I have used:

https://gist.github.com/Cheezykins/7f6f3238c61dc2a8feca30169f988816

I am still experiencing the bug, so it looks like the logic in the refresh() method is either incorrect or not clearing itself afterwards somehow.

Any suggestions?

**Comment by Cheezykins** [comment]
This seems to be a race condition somewhere, although I have not identified where. If I put a gap between the requests then it works fine. Same result if I put a sleep within the refresh token middleware.

**Comment by Cheezykins** [comment]
Ok, I narrowed it down to PayloadFactory::JTI only using second resolution in the hash and not using any form of random element. Meaning two requests in quick succession could have the same JTI.

I went to do a PR and note this is already [fixed in develop](https://github.com/tymondesigns/jwt-auth/blob/develop/src/Factory.php#L210) and the alphas of 1.0.0

So I was wondering if there are any ETAs on 1.0.0 coming out of alpha?
